README

               sapnwrfc - SAP Netweaver RFC support for Perl


    Copyright (c) 2006 - 2010 Piers Harding.
    All rights reserved.
    This program is free software; you can redistribute it and/or modify
    it under the terms of either:

    a) the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

    b) the "Artistic License" which comes with this Kit.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
    the GNU General Public License or the Artistic License for more details.

    You should have received a copy of the Artistic License with this
    Kit, in the file named "Artistic".  If not, I'll be glad to provide one.

    You should also have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software Foundation,
    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

    For those of you that choose to use the GNU General Public License,
    my interpretation of the GNU General Public License is that no Perl
    script falls under the terms of the GPL unless you explicitly put
    said script under the terms of the GPL yourself.  Furthermore, any
    object code linked with perl does not automatically fall under the
    terms of the GPL, provided such object code only adds definitions
    of subroutines and variables, and does not otherwise impair the
    resulting interpreter from executing any standard Perl script.  I
    consider linking in C subroutines in this manner to be the moral
    equivalent of defining subroutines in the Perl language itself.  You
    may sell such an object file as proprietary provided that you provide
    or offer to provide the Perl source, as specified by the GNU General
    Public License.  (This is merely an alternate way of specifying input
    to the program.)  You may also sell a binary produced by the dumping of
    a running Perl script that belongs to you, provided that you provide or
    offer to provide the Perl source as specified by the GPL.  (The
    fact that a Perl interpreter and your code are in the same binary file
    is, in this case, a form of mere aggregation.)  This is my interpretation
    of the GPL.  If you still have concerns or difficulties understanding
    my intent, feel free to contact me.  Of course, the Artistic License
    spells all this out for your protection, so you may prefer to use that.


    One last thing - The SAP NW RFC SDK is a prerequisite for this perl package
    to work.  The NW RFC SDK is proprietary software owned by SAP AG.  For this
    reason I will not supply ANY component part of the RFCSDK with this perl 
    package, and further more you must obtain it through the normal channels 
    with SAP AG - ie. you must have a licensed SAP R/3 installation at your
    disposal.


****ATTENTION****

This package will NOT work unless you obtain the latest NW RFC SDK for SAP NewWeaver
from SAP - this is package is commonly referred to as "sapnwrfc", and is a complete
rewrite from the previous one called SAP::Rfc.


Installation

In brief, the following should work on most systems:
  perl Makefile.PL
  make
  make test  << provide all your host routing information.
  make install

Help on the build process is available by doing:
  perl Makefile.PL --help

and it says:

The whole trouble with the build process is one 
of determining what your OS is, and which RFC
library needs to be linked in, and where that 
library lives.

To help with this there are two command line
options that you can use to point me in the 
right direction:

(1) --source /usr/sap/rfcsdk
  This tells me where to look for your installed
  RFCSDK.  This must be the directory that contains
  the ./lib and ./include directories of the RFCSDK
  In the include directory must be saprfc.h etc.
  By default, this looks in /usr/sap/rfcsdk

(2) --addlibs ' -lm -ldl -lpthread '
  This tells me what additional libraries need to
  be passed in at link time - this can be anything
  that you would put in LDFLAGS.
  By default this is ' -lm -ldl -lpthread '

YOU MUST MAKE SURE THAT YOU ADD THE LIB PATH TO LD_LIBRARY_PATH 
EG:
export LD_LIBRARY_PATH=/usr/sap/rfcsdk/lib

Example:

perl Makefile.PL --source /opt/rfcsdk --addlibs '-lsomethingwacky'

My personal build script contains:
perl Makefile.PL --source=/home/piers/code/sap/nwrfcsdk

where the SAR file for the NW RFC SDK has been unpacked into 
/home/piers/code/sap.

Help on obtaining the NW RFC SDK can be found in the OSS notes 
on http://service.sap.com/support.

Ulrich updates the notes:
https://service.sap.com/sap/support/notes/1025361  - availability
https://service.sap.com/sap/support/notes/1056696  - compiling help

Please email directly if you have problems, and are willing to help 
with testing (especially for other platforms) => piers@cpan.org.


Cheers - Piers Harding - R/2 R/3 BASIS, IFACE, JAPH, whatever ...

email: piers@cpan.org




Changes
Fri Jun  1 17:00:23 BST 2007
  * added in nested table support

Fri Jul 13 06:30:32 BST 2007
  * added a bit more documentation

Thu Jun 26 18:59:57 NZST 2008
  * removed connection attributes that are nolonger supported

Tue Oct 14 06:09:26 NZDT 2008
  * Fixed test plans - thanks Shane Boyce

Mon Dec  8 07:02:31 NZDT 2008
  * fix compile flags - thanks Jens Rehsack.

Fri Dec 26 10:57:43 NZDT 2008
  * Fixes for registered RFC work thanks to Matthias Flury

Mon Jun 22 16:51:14 NZST 2009
  * Fixes on connection parameters
  * sapnwrfc-cookbook - a comprehensive set of examples, and discussions there in
  * Better installation instructions
  * revamped RFC server code
  * banishment of RFC client memory leaks
  * thanks to ce.com.au (Adam Kennedy) for all assistance in making these improvements possible.
  
 Tue Jun 23 17:00:52 NZST 2009
   * further banishment of memory leaks when using tables in RFC client calls
 
 Sat Jun 27 17:55:17 NZST 2009
   * very slow memory leak with respect to referencing table entries on RFC client calls.

Thu Sep 10 07:47:40 NZST 2009
   * bumped the version and removed the dependency on 5.010

Sat Aug 13 05:23:27 NZST 2011
   * fixed broken link (wrong format) in cookbook

Fri Jun  1 09:24:06 NZST 2012
   * fix 4byte chars

Wed Jul 25 09:47:09 NZST 2012
   * lots of fixes supplied by stefan and Marcus - thanks guys

Sat Aug 18 08:52:28 NZST 2012
   * more hpux build fixes supplied by stefan

Fri Aug 31 05:42:19 NZST 2012
   * Fix string length detection - thanks Helmut Schröder



lib/SAPNW/Base.pm
package SAPNW::Base;

=pod

    Copyright (c) 2006 - 2010 Piers Harding.
        All rights reserved.

=cut

  use strict;
  require 5.008;
  use Data::Dumper;

  use vars qw($VERSION $DEBUG $SAPNW_RFC_CONFIG);
  $VERSION = '0.36';

  use constant RFCIMPORT     => 1;
  use constant RFCEXPORT     => 2;
  use constant RFCCHANGING   => 3;
  use constant RFCTABLES     => 7;

  use constant RFCTYPE_CHAR  => 0;
  use constant RFCTYPE_DATE  => 1;
  use constant RFCTYPE_BCD   => 2;
  use constant RFCTYPE_TIME  => 3;
  use constant RFCTYPE_BYTE  => 4;
  use constant RFCTYPE_TABLE => 5;
  use constant RFCTYPE_NUM   => 6;
  use constant RFCTYPE_FLOAT => 7;
  use constant RFCTYPE_INT   => 8;
  use constant RFCTYPE_INT2  => 9;
  use constant RFCTYPE_INT1  => 10;
  use constant RFCTYPE_NULL  => 14;
  use constant RFCTYPE_STRUCTURE  => 17;
  use constant RFCTYPE_DECF16  => 23;
  use constant RFCTYPE_DECF34  => 24;
  use constant RFCTYPE_XMLDATA => 28;
  use constant RFCTYPE_STRING  => 29;
  use constant RFCTYPE_XSTRING => 30;
  use constant RFCTYPE_EXCEPTION => 98;

  use constant RFC_OK => 0;
  use constant RFC_COMMUNICATION_FAILURE => 1;
  use constant RFC_LOGON_FAILURE => 2;
  use constant RFC_ABAP_RUNTIME_FAILURE => 3;
  use constant RFC_ABAP_MESSAGE => 4;
  use constant RFC_ABAP_EXCEPTION => 5;
  use constant RFC_CLOSED => 6;
  use constant RFC_CANCELED => 7;
  use constant RFC_TIMEOUT => 8;
  use constant RFC_MEMORY_INSUFFICIENT => 9;
  use constant RFC_VERSION_MISMATCH => 10;
  use constant RFC_INVALID_PROTOCOL => 11;
  use constant RFC_SERIALIZATION_FAILURE => 12;
  use constant RFC_INVALID_HANDLE => 13;
  use constant RFC_RETRY => 14;
  use constant RFC_EXTERNAL_FAILURE => 15;
  use constant RFC_EXECUTED => 16;
  use constant RFC_NOT_FOUND => 17;
  use constant RFC_NOT_SUPPORTED => 18;
  use constant RFC_ILLEGAL_STATE => 19;
  use constant RFC_INVALID_PARAMETER => 20;
  use constant RFC_CODEPAGE_CONVERSION_FAILURE => 21;
  use constant RFC_CONVERSION_FAILURE => 22;
  use constant RFC_BUFFER_TOO_SMALL => 23;
  use constant RFC_TABLE_MOVE_BOF => 24;
  use constant RFC_TABLE_MOVE_EOF => 25;


  # Export useful tools
  my @export_ok = qw(
     debug
   RFCIMPORT
   RFCEXPORT
   RFCCHANGING
   RFCTABLES
   RFCTYPE_CHAR
   RFCTYPE_DATE
   RFCTYPE_BCD
   RFCTYPE_TIME
   RFCTYPE_BYTE
   RFCTYPE_TABLE
   RFCTYPE_NUM
   RFCTYPE_FLOAT
   RFCTYPE_INT
   RFCTYPE_INT2
   RFCTYPE_INT1
   RFCTYPE_NULL
   RFCTYPE_STRUCTURE
   RFCTYPE_DECF16
   RFCTYPE_DECF34
   RFCTYPE_XMLDATA
   RFCTYPE_STRING
   RFCTYPE_XSTRING
   RFCTYPE_EXCEPTION
   RFC_OK
   RFC_COMMUNICATION_FAILURE
   RFC_LOGON_FAILURE
   RFC_ABAP_RUNTIME_FAILURE
   RFC_ABAP_MESSAGE
   RFC_ABAP_EXCEPTION
   RFC_CLOSED
   RFC_CANCELED
   RFC_TIMEOUT
   RFC_MEMORY_INSUFFICIENT
   RFC_VERSION_MISMATCH
   RFC_INVALID_PROTOCOL
   RFC_SERIALIZATION_FAILURE
   RFC_INVALID_HANDLE
   RFC_RETRY
   RFC_EXTERNAL_FAILURE
   RFC_EXECUTED
   RFC_NOT_FOUND
   RFC_NOT_SUPPORTED
   RFC_ILLEGAL_STATE
   RFC_INVALID_PARAMETER
   RFC_CODEPAGE_CONVERSION_FAILURE
   RFC_CONVERSION_FAILURE
   RFC_BUFFER_TOO_SMALL
   RFC_TABLE_MOVE_BOF
   RFC_TABLE_MOVE_EOF
     Dumper
    );

  sub import {
    my ( $caller ) = caller;
    no strict 'refs';
    foreach my $sub ( @export_ok ){
      *{"${caller}::${sub}"} = \&{$sub};
    }
  }

  sub debug {
      return unless $DEBUG;
        print STDERR scalar localtime() . " - ". caller(), ":> " , @_, "\n";
    }


1;
